### PR TITLE
convertPartialPathsToAbsolute optimization

### PR DESCRIPTION
**Pull request by jjuzna**
Added caching for glob search so that filesystem operations are done only once per watch not once per section.

### CODE REVIEW

File: src/coverage-system/sectionfinder.ts [hunk 1]
```diff
@@ -33,10 +32,13 @@ export class SectionFinder {
         sections: Map<string, Section>,
     ): Section | undefined {
         const sectionsArray = Array.from(sections.values());
-        const fileName = textEditor.document.fileName;
+        const res = this.calculateEditorData(textEditor);
+        if (res === undefined) { return ; }
```
Comment (ryanluker) [inline src/coverage-system/sectionfinder.ts]: minor code styling, could we use `if (!res) { return ; }` instead to better match the rest of the file?

File: src/coverage-system/sectionfinder.ts [hunk 2]
```diff
@@ -101,53 +93,54 @@ export class SectionFinder {
     }
 
     /**
-     * Takes paths and tries to make them absolute
-     * based on currently open workspaceFolders
-     * @param path potential partial path to be converted
+     * Calculates relativePath and workspaceFolder for given TextEditor instance
```
Comment (ryanluker) [inline src/coverage-system/sectionfinder.ts]: very nice! this should hopefully work well in the multiple workspace scenario but did you check that? https://code.visualstudio.com/docs/editor/multi-root-workspaces

File: src/coverage-system/sectionfinder.ts [hunk 3]
```diff
@@ -101,53 +93,54 @@ export class SectionFinder {
...
+        const workspaceFolder = workspace.getWorkspaceFolder(editorFileUri);
+        if (!workspaceFolder) { return; } // file is not in workspace - skip it
+        const workspaceFsPath = workspaceFolder.uri.fsPath;
+        const editorFileAbs = normalizeFileName(fileName);
+        const workspaceFile = normalizeFileName(workspaceFsPath);
+        const editorFileRelative = editorFileAbs.substring(workspaceFile.length);
+        const workspaceFolderName = workspaceFolder.name;
+        return { relativePath: editorFileRelative, workspaceFolder: workspaceFolderName};
+    }
+    /**
```
Comment (ryanluker) [inline src/coverage-system/sectionfinder.ts]: minor code styling, we just need a return between the two functions

File: test/helpers.test.ts [hunk 4]
```diff
@@ -42,25 +43,32 @@ suite("helper Tests", function() {
...
-            const testName = `file1 = '${parameters.file1}' and file2 = '${parameters.file2}' and folder = '${parameters.folder}' -> '${parameters.expected}'`;
-            const isEqual = areFilesRelativeEquals(parameters.file1, parameters.file2, parameters.folder);
-            assert.equal(isEqual, parameters.expected, testName);
+            const testName = `path = '${parameters.path}' -> '${parameters.expected}'`;
+            assert.equal(isPathAbsolute(parameters.path), parameters.expected, testName);
+        });
+        return done();
+    });
+
+    test("Should convert path to searchable properly @unit", function(done) {
```
Comment (ryanluker) [inline test/helpers.test.ts]: Nice test additions thanks for adding these!

File: src/coverage-system/sectionfinder.ts (hunk 2 above)
Comment (jjuzna) [reply src/coverage-system/sectionfinder.ts]: I haven't test it - but I am relaying on `workspace.getWorkspaceFolder` to work as expected. According to [Adopting-Multi-Root-Workspace-APIs](https://github.com/Microsoft/vscode/wiki/Adopting-Multi-Root-Workspace-APIs) it should be OK.

File: src/coverage-system/sectionfinder.ts (hunk 3 above)
Comment (jjuzna) [reply src/coverage-system/sectionfinder.ts]: Can you please provide me with example - not really sure what to change here.

File: src/coverage-system/sectionfinder.ts (hunk 1 above)
Comment (jjuzna) [reply src/coverage-system/sectionfinder.ts]: done

File: src/coverage-system/sectionfinder.ts (hunk 3 above)
Comment (ryanluker) [reply src/coverage-system/sectionfinder.ts]: https://github.com/ryanluker/vscode-coverage-gutters/pull/230/commits/9f58903e20264ff3ff7820542fc7bf4480eb4141 added it for you.

**Comment by ryanluker** [comment]
@jjuzna I will try to get the QA done later this week and merge it in once done.

One thought I had was around if the glob search is even useful anymore (I guess for cases where the filename doesn't have a workspace in it)?

I will confirm the current glob search situation / scenario when I complete the QA.

File: src/coverage-system/sectionfinder.ts (hunk 3 above)
Comment (jjuzna) [reply src/coverage-system/sectionfinder.ts]: Lost in translation  - return equals newline :)

**Comment by jjuzna** [comment]
> 
> 
> @jjuzna I will try to get the QA done later this week and merge it in once done.
> 
> One thought I had was around if the glob search is even useful anymore (I guess for cases where the filename doesn't have a workspace in it)?
> 
> I will confirm the current glob search situation / scenario when I complete the QA.

My first commit reduced the number of glob search to 1, and second one to 0. If editor file is not in any workspace folders the coverage is not shown (it works the same as before).

Glob is now used only to find coverageFiles & reports.

**Comment by ryanluker** [comment]
@jjuzna I checked this out locally and everything looks good! This will go out with the 2.4.1 release.
I have included the extension bundle below encase you want to use it ahead of time.
[vscode-coverage-gutters-2.4.0-letsplay.zip](https://github.com/ryanluker/vscode-coverage-gutters/files/3513325/vscode-coverage-gutters-2.4.0-letsplay.zip)

**Pull request merged**